ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/PerasCertDB/StateMachine.hs

```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Test.Ouroboros.Storage.PerasCertDB.StateMachine (tests) where
+
+import Control.Monad.State
+import Control.Tracer (nullTracer)
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Set as Set
+import Ouroboros.Consensus.Block
+import Ouroboros.Consensus.Peras.Weight (PerasWeightSnapshot)
+import qualified Ouroboros.Consensus.Storage.PerasCertDB as PerasCertDB
+import Ouroboros.Consensus.Storage.PerasCertDB.API (AddPerasCertResult (..), PerasCertDB)
+import Ouroboros.Consensus.Util.IOLike
+import Ouroboros.Consensus.Util.STM
+import qualified Test.Ouroboros.Storage.PerasCertDB.Model as Model
+import Test.QuickCheck
+import qualified Test.QuickCheck.Monadic as QC
+import Test.QuickCheck.StateModel
+import Test.Tasty
+import Test.Tasty.QuickCheck
+import Test.Util.TestBlock (TestBlock, TestHash (..))
+import Test.Util.TestEnv (adjustQuickCheckTests)
+
+tests :: TestTree
+tests =
+  testGroup
+    "PerasCertDB"
+    [ adjustQuickCheckTests (* 100) $ testProperty "q-d" $ prop_qd
+    ]
+
+prop_qd :: Actions Model -> Property
+prop_qd actions = QC.monadic f $ property () <$ runActions actions
+ where
+  f :: StateT (PerasCertDB IO TestBlock) IO Property -> Property
+  f = ioProperty . flip evalStateT (error "unreachable")
+
+newtype Model = Model (Model.Model TestBlock) deriving (Show, Generic)
+
+instance StateModel Model where
+  data Action Model a where
+    OpenDB :: Action Model ()
+    CloseDB :: Action Model ()
+    AddCert :: ValidatedPerasCert TestBlock -> Action Model AddPerasCertResult
+    GetWeightSnapshot :: Action Model (PerasWeightSnapshot TestBlock)
+    GarbageCollect :: SlotNo -> Action Model ()
+
+  arbitraryAction _ (Model model)
+    | model.open =
+        frequency
+          [ (1, pure $ Some CloseDB)
+          , (20, Some <$> genAddCert)
+          , (20, pure $ Some GetWeightSnapshot)
+          , (5, Some . GarbageCollect . SlotNo <$> arbitrary)
+          ]
+    | otherwise = pure $ Some OpenDB
+   where
+    genAddCert = do
+      roundNo <- PerasRoundNo <$> arbitrary
```

Comment:
By using `arbitrary` on this data type, aren't we generating few cases in which there's a collision in `roundNo`